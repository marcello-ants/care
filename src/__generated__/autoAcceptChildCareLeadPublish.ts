/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { AutoAcceptChildCareLeadPublishInput } from "./globalTypes";

// ====================================================
// GraphQL mutation operation: autoAcceptChildCareLeadPublish
// ====================================================

export interface autoAcceptChildCareLeadPublish_autoAcceptChildCareLeadPublish {
  __typename: "AutoAcceptChildCareLeadPublishResponse";
  /**
   * Batch id of creation lead operation
   */
  batchId: string | null;
}

export interface autoAcceptChildCareLeadPublish {
  /**
   * Creates an auto-accept lead
   */
  autoAcceptChildCareLeadPublish: autoAcceptChildCareLeadPublish_autoAcceptChildCareLeadPublish;
}

export interface autoAcceptChildCareLeadPublishVariables {
  autoAcceptChildCareLeadPublishInput: AutoAcceptChildCareLeadPublishInput;
}
