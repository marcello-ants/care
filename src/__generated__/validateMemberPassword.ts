/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: validateMemberPassword
// ====================================================

export interface validateMemberPassword_validateMemberPassword_errors {
  __typename: "MemberPasswordInvalid" | "MemberPasswordInvalidLength" | "MemberPasswordInvalidSequence" | "MemberPasswordInvalidTerms";
  /**
   * Summary of the error.
   */
  message: string | null;
}

export interface validateMemberPassword_validateMemberPassword {
  __typename: "ValidateMemberPasswordPayload";
  /**
   * Array of password validation errors.  The array will be empty when no validation errors were found.
   */
  errors: validateMemberPassword_validateMemberPassword_errors[];
}

export interface validateMemberPassword {
  /**
   * Validates a member's password.
   */
  validateMemberPassword: validateMemberPassword_validateMemberPassword;
}

export interface validateMemberPasswordVariables {
  password: string;
}
