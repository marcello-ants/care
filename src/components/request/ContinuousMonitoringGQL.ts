import { gql } from '@apollo/client';

export const GET_CONTINUOUS_MONITORING_ENROLLMENT_STATUS = gql`
  query ContinuousMonitoringEnrollmentStatus($memberId: ID!) {
    continuousMonitoringEnrollmentStatus(memberId: $memberId) {
      ... on ContinuousMonitoringEnrollmentStatusSuccess {
        needsConsent
        consentMessagesList {
          consentMessageId
          message
        }
      }
    }
  }
`;

export const ENROLL_IN_CONTINUOUS_MONITORING = gql`
  mutation EnrollInContinuousMonitoring($input: EnrollInContinuousMonitoringInput!) {
    enrollInContinuousMonitoring(input: $input) {
      ... on EnrollInContinuousMonitoringSuccess {
        success
      }
    }
  }
`;
