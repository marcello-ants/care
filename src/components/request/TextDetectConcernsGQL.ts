import { gql } from '@apollo/client';

// apollo client breaks if this query uses union type inside fragments and silently drops detectedConcerns from results
const TEXT_DETECT_CONCERNS = gql`
  query textAnalyzerDetectConcerns(
    $bio: String!
    $headline: String!
    $textCreator: TextAnalyzerDetectConcernsCreator!
    $validationId: String!
  ) {
    bio: textAnalyzerDetectConcerns(
      textToAnalyze: { text: $bio }
      textCreator: $textCreator
      validationId: $validationId
    ) {
      ... on TextAnalyzerDetectConcernsSuccess {
        __typename
        detectedConcerns {
          endIndex
          startIndex
          text
          type
        }
        validationId
      }
      ... on TextAnalyzerDetectConcernsError {
        __typename
        message
      }
    }
    headline: textAnalyzerDetectConcerns(
      textToAnalyze: { text: $headline }
      textCreator: $textCreator
      validationId: $validationId
    ) {
      ... on TextAnalyzerDetectConcernsSuccess {
        __typename
        detectedConcerns {
          endIndex
          startIndex
          text
          type
        }
        validationId
      }
      ... on TextAnalyzerDetectConcernsError {
        __typename
        message
      }
    }
  }
`;

export { TEXT_DETECT_CONCERNS };
