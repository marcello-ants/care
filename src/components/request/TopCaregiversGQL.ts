import { gql } from '@apollo/client';

const GET_TOP_CAREGIVERS = gql`
  query getTopCaregivers(
    $zipcode: String!
    $serviceID: ServiceType!
    $numResults: Int!
    $hourlyRate: JobRateInput
    $hasCareCheck: Boolean
    $qualities: [SeniorCareProviderQuality!]
    $services: [SeniorCareServiceProvidedType!]
    $maxDistanceFromSeeker: DistanceInput
    $hasApprovedActivePhoto: Boolean
    $enableFuzzySearch: Boolean
    $daysSinceLastActive: Int
    $isActiveAccount: Boolean
    $minimumAvgReviewRating: Float
    $fullTime: Boolean
  ) {
    topCaregivers: getTopCaregivers(
      zipcode: $zipcode
      numResults: $numResults
      serviceType: $serviceID
      hourlyRate: $hourlyRate
      hasCareCheck: $hasCareCheck
      qualities: $qualities
      services: $services
      maxDistanceFromSeeker: $maxDistanceFromSeeker
      hasApprovedActivePhoto: $hasApprovedActivePhoto
      enableFuzzySearch: $enableFuzzySearch
      daysSinceLastActive: $daysSinceLastActive
      isActiveAccount: $isActiveAccount
      minimumAvgReviewRating: $minimumAvgReviewRating
      fullTime: $fullTime
    ) {
      caregiver {
        profileURL
        avgReviewRating
        numberOfReviews
        yearsOfExperience
        hasCareCheck
        responseTime
        member {
          address {
            city
            state
          }
          displayName
          firstName
          id
          imageURL
          legacyId
        }
        profiles {
          commonCaregiverProfile {
            id
          }
          childCareCaregiverProfile {
            bio {
              experienceSummary
            }
            qualities {
              certifiedNursingAssistant
              certifiedRegisterNurse
              certifiedTeacher
              childDevelopmentAssociate
              comfortableWithPets
              cprTrained
              crn
              doesNotSmoke
              doula
              earlyChildDevelopmentCoursework
              earlyChildhoodEducation
              firstAidTraining
              nafccCertified
              ownTransportation
              specialNeedsCare
              trustlineCertifiedCalifornia
            }
            supportedServices {
              carpooling
              craftAssistance
              errands
              groceryShopping
              laundryAssistance
              lightHousekeeping
              mealPreparation
              swimmingSupervision
              travel
            }
            payRange {
              hourlyRateFrom {
                amount
                currencyCode
              }
              hourlyRateTo {
                amount
                currencyCode
              }
            }
          }
        }
        seniorCareProviderProfile {
          bio
          services
          payRange {
            hourlyRateFrom {
              amount
              currencyCode
            }
            hourlyRateTo {
              amount
              currencyCode
            }
          }
          qualities
        }
        signUpDate
      }
      distanceFromRequestZip {
        unit
        value
      }
    }
  }
`;

export default GET_TOP_CAREGIVERS;
