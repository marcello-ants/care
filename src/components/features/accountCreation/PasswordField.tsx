import React, { useCallback, useRef, useState } from 'react';
import { useApolloClient } from '@apollo/client';
import { useField, useFormikContext } from 'formik';
import { debounce } from 'lodash-es';
import { InlineTextField } from '@care/react-component-lib';
import {
  validateMemberPassword,
  // eslint-disable-next-line camelcase
  validateMemberPassword_validateMemberPassword_errors,
  validateMemberPasswordVariables,
} from '@/__generated__/validateMemberPassword';
import { VALIDATE_MEMBER_PASSWORD } from '@/components/request/GQL';
import { SKIP_AUTH_CONTEXT_KEY } from '@/constants';
import { TextFieldProps } from '@material-ui/core';
import PasswordRequirements from '../PasswordRequirements/PasswordRequirements';

const getTooShortErrorMsg = (minLength: number) =>
  `Password must be at least ${minLength} characters.`;

type PasswordFieldProps = {
  id: string;
  name: string;
  label: string;
  defaultHelperText?: string;
  showErrorWhenTouched?: boolean;
  classes?: TextFieldProps['classes'];
  minLength?: number;
  allowEmptyField?: boolean;
  passwordCopyTest?: number;
};

const isLongEnough = (password: string, minLength: number) => password.length >= minLength;

function PasswordField(props: PasswordFieldProps) {
  const {
    id,
    name,
    label,
    defaultHelperText,
    showErrorWhenTouched,
    classes,
    minLength = 6,
    allowEmptyField,
    passwordCopyTest,
  } = props;
  const ref = useRef<{ password?: string; errorMessage?: string }>({});
  const apolloClient = useApolloClient();

  const debouncedValidation = useCallback(
    debounce(async (password: string, resolve: (msg: string | void) => void) => {
      // eslint-disable-next-line camelcase
      let responseError: validateMemberPassword_validateMemberPassword_errors | null;
      try {
        const { data } = await apolloClient.query<
          validateMemberPassword,
          validateMemberPasswordVariables
        >({
          query: VALIDATE_MEMBER_PASSWORD,
          variables: { password },
          context: { [SKIP_AUTH_CONTEXT_KEY]: true },
        });
        responseError = data?.validateMemberPassword.errors?.[0];
      } catch (e) {
        responseError = null;
      }

      let errorMessage;

      if (responseError) {
        if (
          responseError.__typename === 'MemberPasswordInvalidTerms' &&
          (password.includes('<') || password.includes('>'))
        ) {
          errorMessage = 'Sorry, please remove any < or > from the password and try again';
        } else {
          errorMessage = responseError.message || 'Validation Failed';
        }
      }

      resolve(errorMessage);
      ref.current = {
        password,
        errorMessage,
      };
    }, 500),
    [apolloClient]
  );

  const validate = useCallback(
    async (value: string) => {
      if (allowEmptyField && !value) {
        return '';
      }

      if (ref.current.password === value) {
        return ref.current.errorMessage;
      }

      if (!isLongEnough(value, minLength)) {
        return getTooShortErrorMsg(minLength);
      }

      return new Promise<string | void>((resolve) => {
        debouncedValidation(value, resolve);
      });
    },
    [debouncedValidation]
  );

  let helperText = defaultHelperText;

  if (passwordCopyTest === 2) {
    helperText = 'Minimum 8 characters';
  } else if (passwordCopyTest === 3) {
    helperText = '';
  }

  const [showPasswordRequirements, setShowPasswordRequirements] = useState(false);

  const [isFocused, setIsFocused] = useState(false);
  const { isValidating } = useFormikContext();
  const [field, meta] = useField({ name, validate });

  // server validation errors will be immediately shown, but client-side validation shows when the field is blurred
  const showError = Boolean(
    meta.error && // there's an error to show
      !isValidating && // formik is done validating
      (showErrorWhenTouched ? meta.touched : field.value) && // something was entered OR it was toched (focus & blur)
      (!isFocused || isLongEnough(field.value, minLength)) // only show server-side validations when in focus
  );

  function handleFocus() {
    setIsFocused(true);
    setShowPasswordRequirements(passwordCopyTest === 3);
  }

  function handleBlur(e: React.FocusEvent) {
    setIsFocused(false);
    field.onBlur(e);
  }

  return (
    <>
      <InlineTextField
        type="password"
        classes={classes}
        id={id}
        name={field.name}
        label={label}
        value={field.value}
        error={showError}
        helperText={showError ? meta.error : helperText}
        onChange={field.onChange}
        onBlur={handleBlur}
        onFocus={handleFocus}
      />
      {showPasswordRequirements && <PasswordRequirements />}
    </>
  );
}

PasswordField.defaultProps = {
  showErrorWhenTouched: false,
  classes: undefined,
  minLength: 6,
  allowEmptyField: false,
  defaultHelperText: 'Minimum 6 characters',
  passwordCopyTest: undefined,
};

export default PasswordField;
