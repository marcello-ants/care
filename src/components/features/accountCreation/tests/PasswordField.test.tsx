import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { Formik } from 'formik';

import { VALIDATE_MEMBER_PASSWORD } from '@/components/request/GQL';
// eslint-disable-next-line camelcase
import { validateMemberPassword_validateMemberPassword_errors } from '@/__generated__/validateMemberPassword';
import PasswordField from '../PasswordField';

const passwordValidationMock = (
  password: string,
  // eslint-disable-next-line camelcase
  error?: validateMemberPassword_validateMemberPassword_errors
) => ({
  request: {
    query: VALIDATE_MEMBER_PASSWORD,
    variables: {
      password,
    },
  },
  result: {
    data: {
      validateMemberPassword: { errors: [error] },
    },
  },
});

const passwordValidationErrorMock = {
  request: {
    query: VALIDATE_MEMBER_PASSWORD,
    variables: {
      password: 'blowup',
    },
  },
  error: new Error('ruh roh'),
};

let passwordField: HTMLElement;

function renderComponent(mocks?: ReadonlyArray<MockedResponse>, props?: { [key: string]: any }) {
  const view = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Formik
        initialValues={{
          password: '',
        }}
        validate={(values) => {
          const errors = {};

          if (props?.allowEmptyField && !values.password) {
            // @ts-ignore
            errors.password = '';
          } else if (!values.password) {
            // @ts-ignore
            errors.password = 'Password is required.';
          }

          return errors;
        }}
        onSubmit={() => {}}
        validateOnMount>
        <PasswordField
          id="password"
          name="password"
          label="Password"
          defaultHelperText="This is the helper text"
          {...props}
        />
      </Formik>
    </MockedProvider>
  );

  passwordField = screen.getByLabelText('Password');

  return view;
}

describe('PasswordField', () => {
  it('matches snapshot', () => {
    const view = renderComponent();
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should display a validation error when an invalid password is entered', async () => {
    renderComponent();

    fireEvent.change(passwordField, { target: { value: 'foo' } });

    const validationMsg = await screen.findByText('Password must be at least 6 characters.');
    expect(validationMsg).toBeVisible();
  });

  it('should not display the invalid password message when the field is focused', async () => {
    renderComponent();

    fireEvent.change(passwordField, { target: { value: 'foo' } });

    await screen.findByText('Password must be at least 6 characters.');

    fireEvent.focus(passwordField);

    await waitFor(() =>
      expect(screen.queryByText('Password must be at least 6 characters.')).not.toBeInTheDocument()
    );
  });

  it('should not display the invalid password message when the field is focused and with custom length', async () => {
    renderComponent(undefined, { minLength: 8 });

    fireEvent.change(passwordField, { target: { value: 'foo' } });

    await screen.findByText('Password must be at least 8 characters.');

    fireEvent.focus(passwordField);

    await waitFor(() =>
      expect(screen.queryByText('Password must be at least 6 characters.')).not.toBeInTheDocument()
    );
  });

  it('should display the invalid password message when the field is blurred', async () => {
    renderComponent();

    fireEvent.change(passwordField, { target: { value: 'foo' } });
    fireEvent.focus(passwordField);

    await waitFor(() =>
      expect(screen.queryByText('Password must be at least 6 characters.')).not.toBeInTheDocument()
    );

    fireEvent.blur(passwordField);

    expect(await screen.findByText('Password must be at least 6 characters.')).toBeVisible();
  });

  it('should display server-side validation errors', async () => {
    renderComponent([
      passwordValidationMock('123456', {
        __typename: 'MemberPasswordInvalid',
        message: 'I have the same password on my luggage',
      }),
    ]);

    fireEvent.change(passwordField, { target: { value: '123456' } });

    const validationMsg = await screen.findByText('I have the same password on my luggage');
    expect(validationMsg).toBeVisible();
  });

  it('should display html validation errors', async () => {
    renderComponent([
      passwordValidationMock('12345<6', {
        __typename: 'MemberPasswordInvalidTerms',
        message: 'Server side cryptic message',
      }),
    ]);

    fireEvent.change(passwordField, { target: { value: '12345<6' } });

    const validationMsg = await screen.findByText(/from the password and try again/);
    expect(validationMsg).toBeVisible();
  });

  it('should display "validation failed" when an unexpected response is returned', async () => {
    renderComponent([
      passwordValidationMock('password', {
        __typename: 'MemberPasswordInvalid',
        message: '',
      }),
    ]);

    fireEvent.change(passwordField, { target: { value: 'password' } });

    const validationMsg = await screen.findByText('Validation Failed');
    expect(validationMsg).toBeVisible();
  });

  it('should ignore validation failures', async () => {
    renderComponent([passwordValidationErrorMock]);

    fireEvent.change(passwordField, { target: { value: 'foo' } });

    await screen.findByText('Password must be at least 6 characters.');

    fireEvent.change(passwordField, { target: { value: 'blowup' } });

    await waitFor(() =>
      expect(screen.queryByText('Password must be at least 6 characters.')).not.toBeInTheDocument()
    );

    expect(screen.getByText('This is the helper text')).toBeVisible();
  });

  it('should display password is empty message when the field is blurred and empty', async () => {
    renderComponent([], { showErrorWhenTouched: true });

    fireEvent.focus(passwordField);
    fireEvent.blur(passwordField);

    expect(await screen.findByText('Password is required.')).toBeVisible();
  });
  it('password should be optional', async () => {
    renderComponent([passwordValidationErrorMock], { allowEmptyField: true });

    fireEvent.change(passwordField, { target: { value: 'foo' } });

    await screen.findByText('Password must be at least 6 characters.');

    fireEvent.change(passwordField, { target: { value: '' } });

    await waitFor(() =>
      expect(screen.queryByText('Password must be at least 6 characters.')).not.toBeInTheDocument()
    );
    await waitFor(() =>
      expect(screen.queryByText('Password is required.')).not.toBeInTheDocument()
    );

    expect(screen.getByText('This is the helper text')).toBeVisible();
  });
  it('password copy test with password rules', () => {
    renderComponent([], { passwordCopyTest: 3 });
    fireEvent.focus(passwordField);
    expect(screen.getByText('No more than 2 repeated characters in a row')).toBeVisible();
    fireEvent.blur(passwordField);
    expect(screen.getByText('No more than 2 repeated characters in a row')).toBeVisible();
  });
  it('password copy test with min-8 copy change', async () => {
    renderComponent([], { passwordCopyTest: 2 });
    expect(screen.getByText('Minimum 8 characters')).toBeVisible();
  });
});
