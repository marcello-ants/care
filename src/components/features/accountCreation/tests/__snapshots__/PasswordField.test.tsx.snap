// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordField matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root makeStyles-inline-1"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink="false"
      for="password"
      id="password-label"
    >
      Password
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-describedby="password-helper-text"
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input"
        id="password"
        name="password"
        type="password"
        value=""
      />
    </div>
    <p
      class="MuiFormHelperText-root"
      id="password-helper-text"
    >
      This is the helper text
    </p>
  </div>
</DocumentFragment>
`;
